sdks/python/apache_beam/io/gcp/pubsub_io_perf_test.py

```diff
@@ -0,0 +1,195 @@
+    self.timeout = int(self.pipeline.get_option('timeout'))
+    self.pubsub_name = self.pipeline.get_option('pubsub_name')
+
+  def _setup_pubsub(self):
+    self.pub_client = pubsub.PublisherClient()
+    self.topic_name = self.pub_client.topic_path(
+        self.project_id, self.pubsub_name)
+
+    self.sub_client = pubsub.SubscriberClient()
+    self.read_sub_name = self.sub_client.subscription_path(
+        self.project_id,
+        self.pubsub_name + '_read',
+    )
+    self.read_matcher_sub_name = self.sub_client.subscription_path(
+        self.project_id,
+        self.pubsub_name + '_read_matcher',
+    )
+    self.write_matcher_sub_name = self.sub_client.subscription_path(
+        self.project_id,
+        self.pubsub_name + '_write_matcher',
+    )
+
+  def _setup_pipeline(self, subscription_name):
+    pubsub_msg_verifier = PubSubMessageMatcher(
+        self.project_id,
+        subscription_name,
+        expected_msg_len=self.num_of_messages,
+        timeout=self.timeout,
+        sleep_time=0,
+        max_messages_in_one_pull=1000,
+        pull_timeout=1200)
+
+    extra_opts = {
+        'on_success_matcher': all_of(pubsub_msg_verifier),
+        'wait_until_finish_duration': self.timeout * 1000,
+        'streaming': True,
+        'save_main_session': True
+    }
+
+    args = self.pipeline.get_full_options_as_args(**extra_opts)
+    self.pipeline = TestPipeline(options=PipelineOptions(args))
+
+
+class PubsubWritePerfTest(PubsubIOPerfTest):
+  def __init__(self):
+    super(PubsubWritePerfTest, self).__init__('_write')
+    self._setup_env()
+    self._setup_pubsub()
+    self._setup_pipeline(self.write_matcher_sub_name)
+
+  def test(self):
+    def to_pubsub_message(element):
+      import uuid
+      from apache_beam.io import PubsubMessage
+      return PubsubMessage(
+          data=element[1],
+          attributes={'id': str(uuid.uuid1()).encode('utf-8')},
+      )
+
+    _ = (
+        self.pipeline
+        | 'Create input' >> Read(
+            SyntheticSource(self.parse_synthetic_source_options()))
+        | 'Format to pubsub message in bytes' >> beam.Map(to_pubsub_message)
+        | 'Measure time' >> beam.ParDo(MeasureTime(self.metrics_namespace))
+        | 'Write to Pubsub' >> beam.io.WriteToPubSub(
+            self.topic_name, with_attributes=True, id_label='id'))
+
+  def _setup_pubsub(self):
+    super(PubsubWritePerfTest, self)._setup_pubsub()
+
+    def make_subscription(sub_name):
+      _ = self.sub_client.create_subscription(
+          sub_name,
+          self.topic_name,
+          ack_deadline_seconds=300,
+          timeout=1200,
+      )
+
+    _ = self.pub_client.create_topic(self.topic_name)
+    make_subscription(self.write_matcher_sub_name)
+    make_subscription(self.read_sub_name)
+    make_subscription(self.read_matcher_sub_name)
+
+
+class PubsubReadPerfTest(PubsubIOPerfTest):
+  def __init__(self):
+    super(PubsubReadPerfTest, self).__init__('_read')
```

Comment:
It's totally fine to write all metrics to the same BigQuery table. That's how all IO IT tests in Java SDK works, where we have two types of metric: "read_time" and "write_time". The only thing you have to do is to change the argument that `MeasureTime`'s constructor takes (it's self.metrics_namespace now, which is BigQuery table, but you wanna change that to, say, 'read_time' in PubsubReadPerfTest and 'write_time' in PubsubWritePerfTest)

Reply:
Thanks to that, your changes to the base class will be unnecessary.